File: 20.valid_parentheses.py
# https://leetcode.com/problems/valid-parentheses/description/


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        d = {}
        for x in s:
            if x in d:
                d[x] += 1
            else:
                d[x] = 1
        if (
            d.get("(") != d.get(")")
            or d.get("[") != d.get("]")
            or d.get("{") != d.get("}")
        ):
            return False

        stack = []
        for i, _ in enumerate(s):
            if len(stack) == 0:
                stack.append(i)
                continue

            pi = stack[-1]
            stack.append(i)
            if (
                (s[pi] == "(" and s[i] == ")")
                or (s[pi] == "[" and s[i] == "]")
                or (s[pi] == "{" and s[i] == "}")
            ):
                stack.pop()
                stack.pop()

        return len(stack) == 0


s = Solution()
print(s.isValid("()"))
print(s.isValid("()[]{}"))
print(s.isValid("(]"))
print(s.isValid("([)]"))

File: 20.valid_parentheses.go
// https://leetcode.com/problems/valid-parentheses/description/

package main

import "fmt"

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
	fmt.Println(isValid("([)]"))
}

func isValid(s string) bool {
	if len(s)%2 == 1 {
		return false
	}

	if count('(', s) != count(')', s) ||
		count('[', s) != count(']', s) ||
		count('{', s) != count('}', s) {
		return false
	}

	stack := []int{}
	for i := 0; i < len(s); i++ {
		if len(stack) == 0 {
			stack = append(stack, i)
			continue
		}
		pi := stack[len(stack)-1]
		stack = append(stack, i)
		if (s[pi] == '(' && s[i] == ')') ||
			(s[pi] == '[' && s[i] == ']') ||
			(s[pi] == '{' && s[i] == '}') {
			stack = stack[:len(stack)-2]
		}
	}

	return len(stack) == 0
}

func count(c rune, s string) int {
	count := 0
	for _, v := range s {
		if v == c {
			count++
		}
	}
	return count
}
